cmd/avrogen/main.go

```diff
@@ -75,13 +75,30 @@ func realMain(args []string, out io.Writer) int {
 	}
 	formatted, err := format.Source(buf.Bytes())
 	if err != nil {
-		_, _ = fmt.Fprintf(out, "Error: could format code: %v\n", err)
+		_, _ = fmt.Fprintf(out, "Error: could not format code: %v\n", err)
 		return 3
 	}
-	if err = os.WriteFile(cfg.Out, formatted, 0o600); err != nil {
-		_, _ = fmt.Fprintf(out, "Error: could write file: %v\n", err)
+
+	writer := dumpout
+	if cfg.Out != "" {
+		if writer, err = os.Create(cfg.Out); err != nil {
+			_, _ = fmt.Fprintf(out, "Error: could not create output file: %v\n", err)
+			return 4
+		}
+	}
+
+	if _, err := writer.Write(formatted); err != nil {
+		_, _ = fmt.Fprintf(out, "Error: could not write code: %v\n", err)
```

Comment:
This does not close the file. I think it might be simpler to close the file in a defer after L87?

Reply:
I added a file closing further down below.
But I can make it a defer after opening if you prefer.

Reply:
Using a `defer` would look like this:
```diff
diff --git a/cmd/avrogen/main.go b/cmd/avrogen/main.go
index 3e94091..7ba35a7 100644
--- a/cmd/avrogen/main.go
+++ b/cmd/avrogen/main.go
@@ -81,10 +81,14 @@ func realMain(args []string, out, dumpout io.Writer) int {
 
 	writer := dumpout
 	if cfg.Out != "" {
-		if writer, err = os.Create(cfg.Out); err != nil {
+		file, err := os.Create(cfg.Out)
+		if err != nil {
 			_, _ = fmt.Fprintf(out, "Error: could not create output file: %v\n", err)
 			return 4
 		}
+		defer func() { _ = file.Close() }()
+
+		writer = file
 	}
 
 	if _, err := writer.Write(formatted); err != nil {
@@ -92,13 +96,6 @@ func realMain(args []string, out, dumpout io.Writer) int {
 		return 4
 	}
 
-	if f, ok := writer.(*os.File); ok {
-		if err := f.Close(); err != nil {
-			_, _ = fmt.Fprintf(out, "Error: could not close output file: %v\n", err)
-			return 4
-		}
-	}
-
 	return 0
 }
 ```

Have to use `func() { _ = file.Close() }()` to make golangci-lint happy.
Let me know if you prefer this or the way I had it before.

Reply:
I just realised that unconditionally closing the `writer` if it's a `*io.File` is incorrect behviour because `writer` can be `dumpout` if no filename was provided. In this case, `writer` could be a file (not necessarily stdout) managed by the caller of `realMain` and therefore should not be closed unconditionally.

Therefore, I refactored to close the file only if it was opened by the `realMain` function with a `defer`.